circuits/src/zk_circuits/valid_match_settle_atomic.rs

```diff
@@ -348,9 +348,524 @@ where
+
+    /// Get the maximum amount allowed
+    fn max_amount() -> Amount {
+        (1u128 << AMOUNT_BITS) - 1u128
+    }
+
+    // -----------------------
+    // | Valid Witness Tests |
+    // -----------------------
+
+    /// Tests the `VALID MATCH SETTLE ATOMIC` circuit with a valid witness and
+    /// statement
+    #[test]
+    fn test_valid_match_settle_atomic() {
+        let (witness, statement) = create_witness_statement();
+        assert!(check_constraints(&witness, &statement));
+    }
+
+    // -------------------------
+    // | Matching Engine Tests |
+    // -------------------------
+
+    #[test]
+    fn test_invalid_match_pair() {
+        let (witness, original_statement) = create_witness_statement();
+
+        // Incorrect quote mint
+        let mut statement = original_statement.clone();
+        statement.match_result.quote_mint += 1u8;
+        assert!(!check_constraints(&witness, &statement));
+
+        // Incorrect base mint
+        let mut statement = original_statement;
+        statement.match_result.base_mint += 1u8;
+        assert!(!check_constraints(&witness, &statement));
+    }
+
+    /// Tests the case in which the match goes the wrong direction
+    #[test]
+    fn test_invalid_match_direction() {
+        let (witness, mut statement) = create_witness_statement();
 
-    // TODO: Implement test cases
+        // Incorrect direction
+        statement.match_result.direction = !statement.match_result.direction;
+        assert!(!check_constraints(&witness, &statement));
+    }
+
+    /// Tests the case where the price at which the match executes is invalid
+    #[test]
+    fn test_invalid_match_price() {
+        let (mut witness, statement) = create_witness_statement();
+
+        // Incorrect price, not equal to quote amount / base amount
+        witness.price = witness.price - FixedPoint::from_integer(1);
+        assert!(!check_matching_engine_constraints(&witness, &statement));
+
+        // Invalid price, not representable in `PRICE_BITS` bits
+        let max_representable = BigUint::from(1u8) << PRICE_BITS;
+        let max_repr = biguint_to_scalar(&max_representable);
+        let max_representable_price = FixedPoint::from_repr(max_repr);
+        witness.price = max_representable_price;
+        assert!(!check_matching_engine_constraints(&witness, &statement));
+    }
+
+    /// Test the case in which the quote amount of the match is too high
+    #[test]
+    #[allow(non_snake_case)]
+    fn test_invalid_match__quote_volume_exceeds_bounds() {
+        // Quote amount too high
+        let (mut witness, mut statement) = create_witness_statement_sell_side();
+        let balance = &mut witness.internal_party_balance;
+        let order = &mut witness.internal_party_order;
+
+        // Setup the balance and order to be valid
+        let price_int = 2;
+        let price = FixedPoint::from_integer(price_int);
+        balance.amount = max_amount();
+        order.amount = max_amount() + 1;
```

Comment:
for the confused reader, can we add a quick comment that `order.amount` is constrained to be a valid amount in another circuit, so it being set to `max_amount() + 1` is not what causes constraint insatisfaction here?